selinux/98selinux-microos/selinux-microos-relabel.sh

```diff
@@ -16,64 +16,60 @@ rd_is_selinux_enabled()
-	    ret=$?
-	    info "$out"
+    # If this doesn't exist because e.g. it's not mounted yet due to a bug
+    # (boo#1197309), the exclusion is ignored. If it gets mounted during
+    # the relabel, it gets wrong labels assigned.
+    if ! [ -d "$NEWROOT/var/lib/overlay" ]; then
+	warn "ERROR: /var/lib/overlay doesn't exist - /var not mounted (yet)?"
+	return 1
+    fi
 
-            if [ $ret -eq 0 ]; then
-		#LANG=C /usr/sbin/setenforce 0
-                info "SELinux: mount root read-write and relabel"
-		# Use alternate mount point to prevent overwriting subvolume options (bsc#1186563)
-		ROOT_SELINUX="${NEWROOT}-selinux"
-		mkdir -p "${ROOT_SELINUX}"
-		mount --rbind --make-rslave "${NEWROOT}" "${ROOT_SELINUX}"
-		mount -o remount,rw "${ROOT_SELINUX}"
-                oldrovalue="$(btrfs prop get "${ROOT_SELINUX}" ro | cut -d= -f2)"
-                btrfs prop set "${ROOT_SELINUX}" ro false
-                FORCE=
-		[ -e "${ROOT_SELINUX}"/etc/selinux/.autorelabel ] && FORCE="$(cat "${ROOT_SELINUX}"/etc/selinux/.autorelabel)"
-		LANG=C chroot "${ROOT_SELINUX}" /sbin/restorecon $FORCE -R -e /var/lib/overlay -e /sys -e /dev -e /run /
-                btrfs prop set "${ROOT_SELINUX}" ro "${oldrovalue}"
-		umount -R "${ROOT_SELINUX}"
-            fi
+    ret=0
+    for sysdir in /proc /sys /dev; do
+	if ! mount --rbind "${sysdir}" "${NEWROOT}${sysdir}" ; then
+	    warn "ERROR: mounting ${sysdir} failed!"
+	    ret=1
 	fi
-	for sysdir in /proc /sys /dev; do
-	    if ! umount -R "${NEWROOT}${sysdir}" ; then
-		warn "ERROR: unmounting ${sysdir} failed!"
-		ret=1
-	    fi
-	done
+	# Don't let recursive umounts propagate into the bind source
+	mount --make-rslave "${NEWROOT}${sysdir}"
+    done
+    if [ $ret -eq 0 ]; then
+	#LANG=C /usr/sbin/setenforce 0
+	info "SELinux: mount root read-write and relabel"
+	# Use alternate mount point to prevent overwriting subvolume options (bsc#1186563)
+	ROOT_SELINUX="${NEWROOT}-selinux"
+	mkdir -p "${ROOT_SELINUX}"
+	mount --rbind --make-rslave "${NEWROOT}" "${ROOT_SELINUX}"
+	mount -o remount,rw "${ROOT_SELINUX}"
+	oldrovalue="$(btrfs prop get "${ROOT_SELINUX}" ro | cut -d= -f2)"
+	btrfs prop set "${ROOT_SELINUX}" ro false
+	FORCE=
+	[ -e "${ROOT_SELINUX}"/etc/selinux/.autorelabel ] && FORCE="$(cat "${ROOT_SELINUX}"/etc/selinux/.autorelabel)"
+	. "${ROOT_SELINUX}"/etc/selinux/config
+	info "SELinux: relabel system"
+	# relabel all mounted file systems apart from /var
+	LANG=C chroot "${ROOT_SELINUX}" sh -c 'for mountpoint in $(cut -d" " -f2 /proc/mounts | grep -v "/var$"); do /usr/sbin/setfiles '$FORCE' /etc/selinux/'$SELINUXTYPE'/contexts/files/file_contexts $mountpoint 2>/dev/null; done'
+	# /var is special due to overlayfs
+	LANG=C chroot "${ROOT_SELINUX}" sh -c '/usr/sbin/setfiles '$FORCE' -e /var/lib/overlay /etc/selinux/'$SELINUXTYPE'/contexts/files/file_contexts /var 2>/dev/null'
+	# /etc doesn't get properly labelled due to the overlayfs when
```

Comment:
`/etc` is handled by the first `setfiles` call now

Reply:
Only partially due to the overlayfs setup. You will have unlabelled files in there due to they way `setfiles` works compared to `restorecon`

Reply:
So `/usr/sbin/setfiles /etc/selinux/'$SELINUXTYPE'/contexts/files/file_contexts /etc` does not fully relabel `/etc`?

Why is this not an issue for other mountpoints?

Reply:
because there's no overlayfs mounted on them

Reply:
Does `setfiles` have issues with overlayfs in particular?

Reply:
`setfiles` doesn't enter mointpoints and because of this it doesn't relabel the files that are handled by overlayfs. My technical understanding of overlayfs is slim though

Reply:
> setfiles doesn't enter mointpoints and because of this it doesn't relabel the files that are handled by overlayfs. My technical understanding of overlayfs is slim though

Everything inside `/etc` has the same `st_dev`, so it should be visited properly by just passing `/etc` itself.

Reply:
Do you have an example of a broken `setfiles /etc` result? It appears to work here.

Reply:
Take an unlabeled system, install the policy, boot into the ramdisk (rd.break=pre-pivot) and perform the steps of mounting the root filesystem rw. Then run  
`/usr/sbin/setfiles /etc/selinux/targeted/contexts/files/file_contexts /etc `

For some reason I can't attach the screenshot, here's the link (it takes a while until this is synced):
https://users.suse.com/~jsegitz/grab.jpg

You see there that multiple files (e.g. `/etc/suders`) didn't get labelled

Reply:
I tried the following:

1. Boot into the initrd with `rd.break`
2. Prepare chroot
3. `umount /sysroot/etc`
4. `chroot /sysroot find /var/lib/overlay /etc -exec setfattr -h -x security.selinux {} +`
5. `ls -Z /sysroot/etc` -> everything is unlabelled
6. `systemctl start sysroot-etc.mount`
7. `chroot /sysroot /usr/sbin/setfiles /etc/selinux/targeted/contexts/files/file_contexts /etc`
8. `ls -Z /sysroot/etc` -> everything labelled, no exceptions

So at least with the currently released MicroOS it appears to works fine?

Reply:
where do you break in r`d.break`? I stop in pre-pivot, where the relabeling usually happens and there sysroot-etc is already active, so the `systemctl` doesn't do anything.

Also how do you prepare you chroot? Because you seem to do it differently then what the script (and I) does (operating directly on the `/sysroot` mount, instead of bind mounting it to `/sysroot-selinux`). Maybe that explains the difference

Reply:
btw. I start from a non-SELinux system. So the steps you took to remove the labels aren't necessary for me, but also don't make a difference.

Reply:
I tried running just this command manually, which is pretty much just the original `restorecon` call with explicit `file_contexts` argument:

```
LANG=C chroot "$ROOT_SELINUX" /usr/sbin/setfiles $FORCE -e /var/lib/overlay -e /proc -e /sys -e /dev -e /run "/etc/selinux/$SELINUXTYPE/contexts/files/file_contexts"  $(chroot "$ROOT_SELINUX" cut -d" " -f2 /proc/mounts)
```

and everything was relabelled correctly.

Reply:
so, finally I got it figured out. I thought I had a fully current system, but that wasn't the case due to some repo issues. I was still on kernel 6.1. With 6.2 it works like you see it.

How do we handle this? Older kernels will need the workaround. We could check if /etc is fully labeled and only run the find .. -exec construct if it's not

Reply:
Can we assume that recent enough `setfiles` with the required features is available only with new enough kernel? At least on TW is the case. Then on older versions we keep using `load_policy` + `restorecon`.

Reply:
If we ever update the policy on older SLE Micro releases this problem will pop up there. For TW we're probably good. At least this will be visible quickly in openQA, so let's go with the clean solution now and add a good explanation to the changes file so that it can be quickly found if we run into this

Reply:
I've pushed a version without the special handling for `/etc`. This works for me on current systems.

Reply:
FWICT the single call to setfiles should be enough as well: https://github.com/openSUSE/microos-tools/pull/14#discussion_r1132455164

Reply:
that does not work for me. Can you please try on your side with the policy from `home:jsegitz:branches:security:SELinux`

Reply:
specifically I see a lot of errors, but I can't read them. Then the system reverts to the last snapshot

Reply:
What happens is that because of

```
> zgrep XINO /proc/config.gz 
# CONFIG_OVERLAY_FS_XINO_AUTO is not set
```

`st_dev` of files and directories within `/etc` is not consistent, so it thinks that it consists of several mounpoints which it skips. That is true with all kernel versions though, so maybe it just worked by chance because relevant files were all on the same layer or something like that.

This means the `find /etc -exec setfiles ...` is needed as long as we want to use setfiles on overlayfs without `xino=yes`.

This works here:

```
        LANG=C chroot "$ROOT_SELINUX" /sbin/setfiles $FORCE -e /var/lib/overlay -e /proc -e /sys -e /dev -e /etc "/etc/selinux/${SELINUXTYPE}/contexts/files/file_contexts" $(chroot "$ROOT_SELINUX" cut -d" " -f2 /proc/mounts)
        # On overlayfs, st_dev isn't consistent so setfiles thinks it's a different mountpoint, ignoring it.
        LANG=C chroot "$ROOT_SELINUX" find /etc -exec /sbin/setfiles $FORCE "/etc/selinux/${SELINUXTYPE}/contexts/files/file_contexts" \{\} +
```

So instead of relabelling `/var` separately, relabel `/etc` in a special way. I also removed the unneeded use of `sh -c` and switched to `find -exec +`.

For some reason (at least on SLE Micro 5.4) there is a warning about

```/sbin/setfiles: Could not set context for /.snapshots:  Operation not supported```

That works fine in the started system though. No idea why.

Reply:
I've tested this on my setups and it works. I've updated my PR.

I don't see the error for `/.snapshots`. Quite weird.

So if it's fine for you I would like to get this out now to unblock the policy submit